split_interleave_fastqs.c

```diff
@@ -0,0 +1,313 @@
+    fprintf(stderr,
+            "Error: Expected exactly two positional arguments: <r1_fastq> <r2_fastq>\n");
+    print_usage_and_exit(argv[0]);
+  }
+
+  char *r1_path = argv[optind];
+  char *r2_path = argv[optind + 1];
+
+  // Extract directory and basename from output_prefix
+  // Need separate copies because dirname/basename can modify their inputs
+  char *prefix_copy_dir = strdup(output_prefix);
+  char *prefix_copy_base = strdup(output_prefix);
+  if (!prefix_copy_dir || !prefix_copy_base) {
+    free(prefix_copy_dir);
+    free(prefix_copy_base);
+    perror_and_exit("strdup prefix copy", output_prefix);
+  }
+
+  char *dir_path = dirname(prefix_copy_dir);
+  char *base_name = basename(prefix_copy_base);
+
+  // Initialize resources
+  FILE *f1 = open_fastq_or_die(r1_path);
+  FILE *f2 = open_fastq_or_die(r2_path);
+  FILE *out_chunk_fp = NULL;
+  FastqRecordBuffers r1_buffers = {{NULL}, {0}};
+  FastqRecordBuffers r2_buffers = {{NULL}, {0}};
+  char chunk_filename[MAX_FILENAME_LEN];
+  int chunk_index = 0;
+  long current_chunk_reads = 0;
+  long total_read_pairs = 0;
+  int exit_code = 0;
+
+  fprintf(stderr, "Starting: prefix='%s', max_files=%ld, reads/chunk=%ld\n", output_prefix,
+          max_files, read_pairs_per_chunk);
+
+  // Main processing loop - read and interleave FASTQ records
+  while (true) {
+    bool r1_eof = (read_fastq_record(f1, &r1_buffers, 1) == 0);
+    bool r2_eof = (read_fastq_record(f2, &r2_buffers, 2) == 0);
+
+    if (r1_eof != r2_eof) {
+      fprintf(stderr, "Error: Unequal reads in R1 ('%s') and R2 ('%s').\n", r1_path,
+              r2_path);
+      exit_code = 1;
+      break;
+    }
+
+    if (r1_eof) break;
+
+    if (out_chunk_fp == NULL) {
+      // Wait for a slot and open new chunk file
+      wait_for_file_slot(dir_path, base_name, max_files);
+      snprintf(chunk_filename, MAX_FILENAME_LEN, "%s/%s_chunk%06d.fastq", dir_path,
+               base_name, chunk_index);
+      out_chunk_fp = fopen(chunk_filename, "w");
+      if (!out_chunk_fp) perror_and_exit("fopen output chunk", chunk_filename);
+      fprintf(stderr, "Debug: Opened chunk '%s'\n", chunk_filename);
+      current_chunk_reads = 0;
+    }
+
+    // Write interleaved records
+    for (int i = 0; i < 4; ++i) {
+      if (fputs(r1_buffers.lines[i], out_chunk_fp) == EOF)
+        perror_and_exit("fputs to chunk", chunk_filename);
+    }
+    for (int i = 0; i < 4; ++i) {
+      if (fputs(r2_buffers.lines[i], out_chunk_fp) == EOF)
+        perror_and_exit("fputs to chunk", chunk_filename);
+    }
+
+    total_read_pairs++;
+    current_chunk_reads++;
+
+    // Close chunk when full
+    if (current_chunk_reads >= read_pairs_per_chunk) {
+      if (fclose(out_chunk_fp) != 0) perror_and_exit("fclose chunk", chunk_filename);
+      fprintf(stderr, "Debug: Closed chunk '%s' (%ld pairs)\n", chunk_filename,
+              current_chunk_reads);
+      printf("%s\n", chunk_filename);  // Output chunk filename to stdout, not stderr
```

Comment:
When writing this, it felt most natural to output the full filename of each produced file. But for passing to an xargs wrapper, it might be more convenient to output just the `_chunkNNNNNN.fastq` suffix. I'm open to thoughts and suggestions.

Reply:
Okay I'm going to change this, good talk.